mllib/src/main/scala/org/apache/spark/ml/feature/Binarizer.scala

```diff
@@ -69,66 +83,117 @@ final class Binarizer @Since("1.4.0") (@Since("1.4.0") override val uid: String)
 
+  /** @group setParam */
+  @Since("3.0.0")
+  def setInputCols(value: Array[String]): this.type = set(inputCols, value)
+
+  /** @group setParam */
+  @Since("3.0.0")
+  def setOutputCols(value: Array[String]): this.type = set(outputCols, value)
+
   @Since("2.0.0")
   override def transform(dataset: Dataset[_]): DataFrame = {
     val outputSchema = transformSchema(dataset.schema, logging = true)
-    val schema = dataset.schema
-    val inputType = schema($(inputCol)).dataType
-    val td = $(threshold)
-    val metadata = outputSchema($(outputCol)).metadata
-
-    val binarizerUDF = inputType match {
-      case DoubleType =>
-        udf { in: Double => if (in > td) 1.0 else 0.0 }
-
-      case _: VectorUDT if td >= 0 =>
-        udf { vector: Vector =>
-          val indices = ArrayBuilder.make[Int]
-          val values = ArrayBuilder.make[Double]
-          vector.foreachActive { (index, value) =>
-            if (value > td) {
-              indices += index
-              values +=  1.0
+
+    val (inputColNames, outputColNames, tds) =
+      if (isSet(inputCols)) {
+        if (isSet(thresholds)) {
+          ($(inputCols).toSeq, $(outputCols).toSeq, $(thresholds).toSeq)
+        } else {
+          ($(inputCols).toSeq, $(outputCols).toSeq, Seq.fill($(inputCols).length)($(threshold)))
+        }
+      } else {
+        (Seq($(inputCol)), Seq($(outputCol)), Seq($(threshold)))
+      }
+
+    val ouputCols = inputColNames.zip(tds).map { case (inputColName, td) =>
+      val binarizerUDF = dataset.schema(inputColName).dataType match {
+        case DoubleType =>
+          udf { in: Double => if (in > td) 1.0 else 0.0 }
+
+        case _: VectorUDT if td >= 0 =>
+          udf { vector: Vector =>
+            val indices = ArrayBuilder.make[Int]
+            val values = ArrayBuilder.make[Double]
+            vector.foreachActive { (index, value) =>
+              if (value > td) {
+                indices += index
+                values +=  1.0
+              }
             }
+            Vectors.sparse(vector.size, indices.result(), values.result()).compressed
           }
-          Vectors.sparse(vector.size, indices.result(), values.result()).compressed
-        }
 
-      case _: VectorUDT if td < 0 =>
-        this.logWarning(s"Binarization operations on sparse dataset with negative threshold " +
-          s"$td will build a dense output, so take care when applying to sparse input.")
-        udf { vector: Vector =>
-          val values = Array.fill(vector.size)(1.0)
-          vector.foreachActive { (index, value) =>
-            if (value <= td) {
-              values(index) = 0.0
+        case _: VectorUDT if td < 0 =>
+          this.logWarning(s"Binarization operations on sparse dataset with negative threshold " +
+            s"$td will build a dense output, so take care when applying to sparse input.")
+          udf { vector: Vector =>
+            val values = Array.fill(vector.size)(1.0)
```

Comment:
Nit, while we're here. Is it maybe faster to start with an array of 0, and flip the ones that exceed the threshold? just because the array is already initialized to 0.

Reply:
Oh, if we initialize an array with 0 values, we mast traversal across all elements since the threshold is negativce, implicit 0 will return 0. Then we can not use `foreachActive` to only process active elements.

Reply:
Oh right of course, nevermind.